File: smor_ting_mobile/lib/core/models/service.dart
import 'package:json_annotation/json_annotation.dart';

part 'service.g.dart';

@JsonSerializable()
class ServiceCategory {
  final String id;
  final String name;
  final String description;
  final String icon;
  final String color;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const ServiceCategory({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.color,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ServiceCategory.fromJson(Map<String, dynamic> json) => _$ServiceCategoryFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceCategoryToJson(this);
}

@JsonSerializable()
class Service {
  final String id;
  final String name;
  final String description;
  @JsonKey(name: 'category_id')
  final String categoryId;
  @JsonKey(name: 'provider_id')
  final String providerId;
  final double price;
  final String currency;
  final int duration; // in minutes
  final List<String> images;
  @JsonKey(name: 'is_active')
  final bool isActive;
  final double rating;
  @JsonKey(name: 'review_count')
  final int reviewCount;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const Service({
    required this.id,
    required this.name,
    required this.description,
    required this.categoryId,
    required this.providerId,
    required this.price,
    required this.currency,
    required this.duration,
    required this.images,
    required this.isActive,
    required this.rating,
    required this.reviewCount,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Service.fromJson(Map<String, dynamic> json) => _$ServiceFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceToJson(this);

  String get formattedPrice => '$currency ${price.toStringAsFixed(2)}';
  String get formattedDuration {
    if (duration < 60) {
      return '${duration}min';
    } else {
      final hours = duration ~/ 60;
      final minutes = duration % 60;
      return minutes > 0 ? '${hours}h ${minutes}min' : '${hours}h';
    }
  }
}

@JsonSerializable()
class ServiceProvider {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'business_name')
  final String businessName;
  final String description;
  final int experience; // years
  final List<String> certifications;
  @JsonKey(name: 'service_areas')
  final List<String> serviceAreas;
  @JsonKey(name: 'is_verified')
  final bool isVerified;
  final double rating;
  @JsonKey(name: 'review_count')
  final int reviewCount;
  @JsonKey(name: 'completed_jobs')
  final int completedJobs;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const ServiceProvider({
    required this.id,
    required this.userId,
    required this.businessName,
    required this.description,
    required this.experience,
    required this.certifications,
    required this.serviceAreas,
    required this.isVerified,
    required this.rating,
    required this.reviewCount,
    required this.completedJobs,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ServiceProvider.fromJson(Map<String, dynamic> json) => _$ServiceProviderFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceProviderToJson(this);

  String get experienceText => '$experience year${experience != 1 ? 's' : ''} experience';
}

enum BookingStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('confirmed')
  confirmed,
  @JsonValue('in_progress')
  inProgress,
  @JsonValue('completed')
  completed,
  @JsonValue('cancelled')
  cancelled,
}

@JsonSerializable()
class Booking {
  final String id;
  @JsonKey(name: 'customer_id')
  final String customerId;
  @JsonKey(name: 'provider_id')
  final String providerId;
  @JsonKey(name: 'service_id')
  final String serviceId;
  final BookingStatus status;
  @JsonKey(name: 'scheduled_date')
  final DateTime scheduledDate;
  @JsonKey(name: 'completed_date')
  final DateTime? completedDate;
  final Address address;
  final String notes;
  @JsonKey(name: 'total_amount')
  final double totalAmount;
  final String currency;
  @JsonKey(name: 'payment_status')
  final String paymentStatus;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  const Booking({
    required this.id,
    required this.customerId,
    required this.providerId,
    required this.serviceId,
    required this.status,
    required this.scheduledDate,
    this.completedDate,
    required this.address,
    required this.notes,
    required this.totalAmount,
    required this.currency,
    required this.paymentStatus,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) => _$BookingFromJson(json);
  Map<String, dynamic> toJson() => _$BookingToJson(this);

  String get formattedAmount => '$currency ${totalAmount.toStringAsFixed(2)}';
}

@JsonSerializable()
class Address {
  final String street;
  final String city;
  final String county;
  final String country;
  final double latitude;
  final double longitude;

  const Address({
    required this.street,
    required this.city,
    required this.county,
    required this.country,
    required this.latitude,
    required this.longitude,
  });

  factory Address.fromJson(Map<String, dynamic> json) => _$AddressFromJson(json);
  Map<String, dynamic> toJson() => _$AddressToJson(this);

  String get fullAddress => '$street, $city, $county, $country';
}

@JsonSerializable()
class Review {
  final String id;
  @JsonKey(name: 'booking_id')
  final String bookingId;
  @JsonKey(name: 'customer_id')
  final String customerId;
  @JsonKey(name: 'provider_id')
  final String providerId;
  @JsonKey(name: 'service_id')
  final String serviceId;
  final int rating; // 1-5
  final String comment;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;

  const Review({
    required this.id,
    required this.bookingId,
    required this.customerId,
    required this.providerId,
    required this.serviceId,
    required this.rating,
    required this.comment,
    required this.createdAt,
  });

  factory Review.fromJson(Map<String, dynamic> json) => _$ReviewFromJson(json);
  Map<String, dynamic> toJson() => _$ReviewToJson(this);
}

File: smor_ting_backend/internal/models/service.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ServiceCategory struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Name        string             `json:"name" bson:"name"`
	Description string             `json:"description" bson:"description"`
	Icon        string             `json:"icon" bson:"icon"`
	Color       string             `json:"color" bson:"color"`
	IsActive    bool               `json:"is_active" bson:"is_active"`
	CreatedAt   time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt   time.Time          `json:"updated_at" bson:"updated_at"`
}

type Service struct {
	ID          primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Name        string             `json:"name" bson:"name"`
	Description string             `json:"description" bson:"description"`
	CategoryID  primitive.ObjectID `json:"category_id" bson:"category_id"`
	ProviderID  primitive.ObjectID `json:"provider_id" bson:"provider_id"`
	Price       float64            `json:"price" bson:"price"`
	Currency    string             `json:"currency" bson:"currency"`
	Duration    int                `json:"duration" bson:"duration"` // in minutes
	Images      []string           `json:"images" bson:"images"`
	IsActive    bool               `json:"is_active" bson:"is_active"`
	Rating      float64            `json:"rating" bson:"rating"`
	ReviewCount int                `json:"review_count" bson:"review_count"`
	// Embedded reviews for better performance
	Reviews []Review `json:"reviews,omitempty" bson:"reviews,omitempty"`
	// Location for geospatial queries
	Location Address `json:"location" bson:"location"`
	// Offline-first fields
	LastSyncAt time.Time `json:"last_sync_at" bson:"last_sync_at"`
	Version    int       `json:"version" bson:"version"`
	CreatedAt  time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt  time.Time `json:"updated_at" bson:"updated_at"`
}

type ServiceProvider struct {
	ID             primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	UserID         primitive.ObjectID `json:"user_id" bson:"user_id"`
	BusinessName   string             `json:"business_name" bson:"business_name"`
	Description    string             `json:"description" bson:"description"`
	Experience     int                `json:"experience" bson:"experience"` // years
	Certifications []string           `json:"certifications" bson:"certifications"`
	ServiceAreas   []string           `json:"service_areas" bson:"service_areas"`
	IsVerified     bool               `json:"is_verified" bson:"is_verified"`
	Rating         float64            `json:"rating" bson:"rating"`
	ReviewCount    int                `json:"review_count" bson:"review_count"`
	CompletedJobs  int                `json:"completed_jobs" bson:"completed_jobs"`
	// Embedded services for better performance
	Services []Service `json:"services,omitempty" bson:"services,omitempty"`
	// Offline-first fields
	LastSyncAt time.Time `json:"last_sync_at" bson:"last_sync_at"`
	Version    int       `json:"version" bson:"version"`
	CreatedAt  time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt  time.Time `json:"updated_at" bson:"updated_at"`
}

type BookingStatus string

const (
	BookingPending    BookingStatus = "pending"
	BookingConfirmed  BookingStatus = "confirmed"
	BookingInProgress BookingStatus = "in_progress"
	BookingCompleted  BookingStatus = "completed"
	BookingCancelled  BookingStatus = "cancelled"
)

type Booking struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	CustomerID    primitive.ObjectID `json:"customer_id" bson:"customer_id"`
	ProviderID    primitive.ObjectID `json:"provider_id" bson:"provider_id"`
	ServiceID     primitive.ObjectID `json:"service_id" bson:"service_id"`
	Status        BookingStatus      `json:"status" bson:"status"`
	ScheduledDate time.Time          `json:"scheduled_date" bson:"scheduled_date"`
	CompletedDate *time.Time         `json:"completed_date,omitempty" bson:"completed_date,omitempty"`
	Address       Address            `json:"address" bson:"address"`
	Notes         string             `json:"notes" bson:"notes"`
	TotalAmount   float64            `json:"total_amount" bson:"total_amount"`
	Currency      string             `json:"currency" bson:"currency"`
	PaymentStatus string             `json:"payment_status" bson:"payment_status"`
	// Embedded service details for offline access
	Service Service `json:"service" bson:"service"`
	// Payment information
	Payment Payment `json:"payment" bson:"payment"`
	// Tracking information
	Tracking Tracking `json:"tracking,omitempty" bson:"tracking,omitempty"`
	// Offline-first fields
	LastSyncAt time.Time `json:"last_sync_at" bson:"last_sync_at"`
	Version    int       `json:"version" bson:"version"`
	CreatedAt  time.Time `json:"created_at" bson:"created_at"`
	UpdatedAt  time.Time `json:"updated_at" bson:"updated_at"`
}

type Payment struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Method    string             `json:"method" bson:"method"` // "mobile_money", "card", "wallet"
	Amount    float64            `json:"amount" bson:"amount"`
	Currency  string             `json:"currency" bson:"currency"`
	Status    string             `json:"status" bson:"status"` // "pending", "completed", "failed"
	Reference string             `json:"reference" bson:"reference"`
	Provider  string             `json:"provider" bson:"provider"` // "mpesa", "airtel_money", etc.
	CreatedAt time.Time          `json:"created_at" bson:"created_at"`
	UpdatedAt time.Time          `json:"updated_at" bson:"updated_at"`
}

type Tracking struct {
	Status           string    `json:"status" bson:"status"`
	Location         Address   `json:"location" bson:"location"`
	UpdatedAt        time.Time `json:"updated_at" bson:"updated_at"`
	EstimatedArrival time.Time `json:"estimated_arrival" bson:"estimated_arrival"`
}

type Review struct {
	ID         primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	BookingID  primitive.ObjectID `json:"booking_id" bson:"booking_id"`
	CustomerID primitive.ObjectID `json:"customer_id" bson:"customer_id"`
	ProviderID primitive.ObjectID `json:"provider_id" bson:"provider_id"`
	ServiceID  primitive.ObjectID `json:"service_id" bson:"service_id"`
	Rating     int                `json:"rating" bson:"rating"` // 1-5
	Comment    string             `json:"comment" bson:"comment"`
	// Offline-first fields
	LastSyncAt time.Time `json:"last_sync_at" bson:"last_sync_at"`
	Version    int       `json:"version" bson:"version"`
	CreatedAt  time.Time `json:"created_at" bson:"created_at"`
}
